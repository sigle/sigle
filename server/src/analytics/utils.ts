import { isBefore } from 'date-fns';

// All queries should start at this date maximum.
const STATS_MAX_FROM_DATE = '2022-05-01';

/**
 * Set max date in the past. Fathom data is not correct before this one.
 */
export const maxFathomFromDate = (parsedDateFrom: Date, dateFrom: string) => {
  if (isBefore(parsedDateFrom, new Date(STATS_MAX_FROM_DATE))) {
    return STATS_MAX_FROM_DATE;
  }
  return dateFrom;
};
