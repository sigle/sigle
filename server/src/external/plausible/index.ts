import { config } from '../../config';
import { pages } from './pages';
import { referrers } from './referrers';
import { timeseries } from './timeseries';

export interface PlausibleClientParams {
  apiToken: string;
  siteId: string;
}

export const initPlausibleClient = (clientParams: PlausibleClientParams) => {
  return {
    timeseries: timeseries(clientParams),
    referrers: referrers(clientParams),
    pages: pages(clientParams),
  };
};

export const plausibleClient = initPlausibleClient({
  apiToken: config.PLAUSIBLE_API_TOKEN,
  siteId: config.PLAUSIBLE_SITE_ID,
});
