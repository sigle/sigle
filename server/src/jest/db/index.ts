import { prisma } from '../../prisma';

export * from './user';

export const TestBaseDB = {
  /**
   * After tests are done, clean up the database
   */
  cleanup: async () => {
    await prisma.subscription.deleteMany({});
    await prisma.follows.deleteMany({});
    await prisma.user.deleteMany({});
  },
};
