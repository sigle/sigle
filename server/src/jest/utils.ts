/**
 * Required when mocking Date objects. Jest is creating some issues with fastify.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const fakeTimerConfigDate: any = {
  doNotFake: [
    'hrtime',
    'nextTick',
    'performance',
    'queueMicrotask',
    'requestAnimationFrame',
    'cancelAnimationFrame',
    'requestIdleCallback',
    'cancelIdleCallback',
    'setImmediate',
    'clearImmediate',
    'setInterval',
    'clearInterval',
    'setTimeout',
    'clearTimeout',
  ],
};
