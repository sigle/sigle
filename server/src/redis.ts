import Redis from 'ioredis';

export const redis = new Redis();
