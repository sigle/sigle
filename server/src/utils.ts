// Production list
export const allowedNewsletterUsers = [
  // sigle.btc
  'SP2EVYKET55QH40RAZE5PVZ363QX0X6BSRP4C7H0W',
  // leopradel.btc
  'SP3VCX5NFQ8VCHFS9M6N40ZJNVTRT4HZ62WFH5C4Q',
  // quentin.btc
  'SP24GYRG3M7T0S6FZE9RVVP9PNNZQJQ614650G590',
  // Test account
  'SPRG63HF7QJJ8H7JN0DJQZH2P1FCGYVHPYBXMYGB',
  // jackbinswitch.btc
  'SPQE3J7XMMK0DN0BWJZHGE6B05VDYQRXRMDV734D',
  // whenindoubt.btc
  'SP398XE371G08T84A99TCBD8XKWY3S7VVX6JKJWKY',
  // nftclothingclub.btc
  'SP2ABRYMZ38D5BHDMWPX6V0PKYA733W2T755DKQ72',
  // stxldn.btc
  'SPCMGSQF3ME39XN7W6RV4M21HFRHC2BGJH6S07V7',
  // bigrpic.btc
  'SP2H6HVZK6X3Z8F4PKF284AZJR6FH4H9J4W6KVV8T',
  // morguf.btc
  'SP21CYC2GKWTVK3FHFF4VVJNKVNQDMRY5GQS27XQB',
  // friedger.btc
  'SPN4Y5QPGQA8882ZXW90ADC2DHYXMSTN8VAR8C3X',
  // johnd.btc
  'SP14W78Q821B3HQ3ED30624Z1F13X4JMFZY3N5SK4',
  // behindthescenes.btc
  'SP1MX63HP0YD1TFAR0J6N6VYN3KVED5AF5JHPH1B7',
  // wampastompa.btc
  'SP22W7TM6NG3PJ2XVVND2E06D50K3DDNREBTKGFD3',
  // marthafog.btc
  'SP3J4WEWR42Q5919MR8CBV4X4ZC19SSA0Q7PKHZVG',
  // Expresso
  'SP0FK47T9271MJX3HEFKA1MEPD4X39J9DCGVXF2D',
  // 21milbtc.btc
  'SP3BCZN307DECNR5PRMV6HY4P37AJ9N48JP0VE547',
  // goodkitty.btc
  'SP28RZ1QXMXJXVKRRCR3D7GR5D48XY0NNA9MZWHJB',
  // stablesir.btc
  'SP8ZVD9PQ9VP657P670KBCB5ZHMJM9QMNJNFVBS6',
  // dutchee.btc
  'SP339T41DGKV2ZR8ACAT630V8KWJVSJME2MK5ZXCB',
  // xverseapp.id.stx
  'SPNW0Q9NE7GSFH10RKYWEN68FBQ9V5JFWCN8SK31',
  // liquidium.btc
  'SPYHY9MV6S08YJQVW0R400ADXZBBJ0GM096BMY34',
  // nakops.btc
  'SP2PGSFKQR4H4AJ093MAXMKMEEDHVFVY881F0FF7F',
  // counterfeitcvlt.btc
  'SPWP8QGG60KA0T3KSZ2RC4M8THZTA0G3DZATMJ6V',
  // stxaus.btc
  'SP162W9JMD2CRP9XDF1AJVE2MSTFQNRNPPKCRDA58',
  // joemacmillan.btc
  'SP23ZSH34VFD9WC6FVWYNKJAVME1SN89ZWHHR5NCC',
  // igolle.btc
  'SP3Y808KQW24TXXVBQ4S71Z22KC93QJCTAJS35GH0',
];
if (process.env.NODE_ENV === 'development') {
  // gregogun.btc
  allowedNewsletterUsers.push('SP1F48HCD4SP4HT8BHQPXZ35615764KC80ACNMBDZ');
}

export const generateAvatar = (stxAddress: string) => {
  return `https://source.boringavatars.com/marble/120/${stxAddress}?square&colors=da3a00,fa8158,ffa835,37c391,7c456cf`;
};
