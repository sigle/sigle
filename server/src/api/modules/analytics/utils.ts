import { lookupProfile } from '@stacks/auth';
import { isBefore } from 'date-fns';
import { fetch } from 'undici';
import { config } from '../../../config';
import { migrationStories } from '../../../external/gaia';

// All queries should start at this date maximum.
const STATS_MAX_FROM_DATE = '2022-05-01';

/**
 * Set max date in the past. Fathom data is not correct before this one.
 */
export const maxFathomFromDate = (parsedDateFrom: Date, dateFrom: string) => {
  if (isBefore(parsedDateFrom, new Date(STATS_MAX_FROM_DATE))) {
    return STATS_MAX_FROM_DATE;
  }
  return dateFrom;
};

export const getBucketUrl = async ({ username }: { username: string }) => {
  let userProfile: Record<string, any> | undefined;
  try {
    userProfile = await lookupProfile({ username });
  } catch (error) {
    // This will happen if there is no blockstack user with this name
    if (error.message === 'Name not found') {
      userProfile = undefined;
    } else {
      throw error;
    }
  }

  const bucketUrl: string | undefined =
    userProfile && userProfile.apps && userProfile.apps[config.APP_URL];

  return { profile: userProfile, bucketUrl };
};

export const getPublicStories = async ({
  bucketUrl,
}: {
  bucketUrl: string;
}) => {
  const resPublicStories = await fetch(`${bucketUrl}publicStories.json`);
  // This would happen if the user has not published any stories
  if (resPublicStories.status !== 200) {
    return [];
  }
  const publicStoriesFile = migrationStories(
    (await resPublicStories.json()) as any
  );
  return publicStoriesFile.stories;
};
