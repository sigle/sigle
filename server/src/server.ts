import Fastify, { FastifyServerOptions, FastifyLoggerInstance } from 'fastify';
import FastifyCors from '@fastify/cors';
import FastifyRateLimit from '@fastify/rate-limit';
import FastifyCookie from '@fastify/cookie';
import { Server } from 'http';
import * as Sentry from '@sentry/node';
import { createAnalyticsHistoricalEndpoint } from './api/modules/analytics/historical';
import { createAnalyticsReferrersEndpoint } from './api/modules/analytics/referrers';
import { config } from './config';
import { redis } from './redis';
import { fastifyAuthPlugin } from './api/plugins/auth';
import { createSubscriptionCreatorPlusEndpoint } from './api/modules/subscriptions/creatorPlus';
import { createGetSubscriptionEndpoint } from './api/modules/subscriptions/getSubscription';

export const buildFastifyServer = (
  opts: FastifyServerOptions<Server, FastifyLoggerInstance> = {}
) => {
  const fastify = Fastify(opts);

  /**
   * Cors is disabled for local env.
   * Cors is enabled on prod and allowed only for the APP_URL.
   * Allow the RENDER API to make calls, used to bypass CORS for the health check, in such case origin will be undefined.
   */
  fastify.register(FastifyCors, {
    credentials: true,
    origin: (origin, cb) => {
      if (
        config.NODE_ENV === 'development' ||
        origin === config.APP_URL ||
        origin === undefined
      ) {
        cb(null, true);
        return;
      }
      cb(new Error('Not allowed'), false);
    },
  });

  /**
   * Rate limit is disabled for local env.
   * Max 50 requests per minute.
   * We increase the limit when we are running tests.
   */
  fastify.register(FastifyRateLimit, {
    max: config.NODE_ENV === 'test' ? 1000 : 50,
    timeWindow: 60000,
    redis,
  });

  /**
   * Parse the cookies sent by the web app for authentication.
   */
  fastify.register(FastifyCookie, {
    secret: config.NEXTAUTH_SECRET,
  });

  /**
   * Authentication
   */
  fastify.register(fastifyAuthPlugin);

  /**
   * Catch and report errors with Sentry.
   * We attach some content to make it easier to debug.
   */
  fastify.setErrorHandler(async (error, request, reply) => {
    // We don't report rate-limit errors to Sentry.
    if (reply.statusCode === 429) {
      reply.send(error);
      return;
    }

    Sentry.withScope((scope) => {
      scope.setLevel(Sentry.Severity.Error);
      scope.setTag('path', request.url);
      scope.setExtra('headers', request.headers);
      if (request.address) {
        scope.setUser({ stacksAddress: request.address });
      }
      if (
        request.headers['content-type'] === 'application/json' &&
        request.body
      ) {
        scope.setExtra('body', request.body);
      }

      const sentryId = Sentry.captureException(error);
      // Also log to the console in case it's not reported to sentry
      console.error(sentryId, error);
      reply.status(500).send({
        error: 'Internal server error',
        message: 'Something went wrong please try again after some time',
        errorId: sentryId,
      });
    });
  });

  fastify.get('/health', (_, res) => {
    res.send({
      success: true,
    });
  });

  /**
   * All the protected routes must be placed there.
   */
  fastify.after(() => {
    /**
     * Subscriptions routes
     */
    createSubscriptionCreatorPlusEndpoint(fastify);
    createGetSubscriptionEndpoint(fastify);

    /**
     * Analytics routes
     */
    createAnalyticsReferrersEndpoint(fastify);
    createAnalyticsHistoricalEndpoint(fastify);
  });

  return fastify;
};
