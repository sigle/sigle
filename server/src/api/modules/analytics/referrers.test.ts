/* eslint-disable @typescript-eslint/no-explicit-any */
import { FastifyInstance } from 'fastify';
import { plausibleClient } from '../../../external/plausible';
import { fakeTimerConfigDate } from '../../../jest/utils';
import { TestBaseDB, TestDBUser } from '../../../jest/db';
import { prisma } from '../../../prisma';
import { redis } from '../../../redis';
import { buildFastifyServer } from '../../../server';

jest.mock('../../../external/plausible');

let server: FastifyInstance;

beforeAll(() => {
  server = buildFastifyServer();
});

afterAll(async () => {
  await TestBaseDB.cleanup();
  await redis.quit();
  await prisma.$disconnect();
});

beforeEach(async () => {
  await prisma.subscription.deleteMany({});
  await prisma.user.deleteMany({});
});

beforeEach(() => {
  (plausibleClient.referrers as jest.Mock).mockReset();
});

it('Respond with referrers', async () => {
  const stacksAddress = 'SP2BKHGRV8H2YDJK16FP5VASYFDGYN4BPTNFWDKYJ';
  await TestDBUser.seedUserWithSubscription({ stacksAddress });

  // Set a fake timer so we can verify the end date
  jest
    .useFakeTimers(fakeTimerConfigDate)
    .setSystemTime(new Date('2022-04-04 12:00:00 UTC').getTime());
  (plausibleClient.referrers as jest.Mock).mockResolvedValueOnce([
    {
      pageviews: '1',
      source: 'https://stacks.co',
    },
    {
      pageviews: '10',
      source: 'Direct / None',
    },
  ]);

  const response = await server.inject({
    method: 'GET',
    url: '/api/analytics/referrers?dateFrom=2022-03-15',
    cookies: {
      'next-auth.session-token': stacksAddress,
    },
  });

  expect(response.statusCode).toBe(200);
  expect(response.json()).toMatchSnapshot();
  expect(plausibleClient.referrers).toBeCalledWith({
    dateFrom: '2022-05-01',
    dateTo: '2022-04-04',
    paths: expect.any(Array),
  });
});

it('Respond with referrers for one story', async () => {
  const stacksAddress = 'SP2BKHGRV8H2YDJK16FP5VASYFDGYN4BPTNFWDKYJ';
  await TestDBUser.seedUserWithSubscription({ stacksAddress });

  // Set a fake timer so we can verify the end date
  jest
    .useFakeTimers(fakeTimerConfigDate)
    .setSystemTime(new Date('2022-04-04 12:00:00 UTC').getTime());
  (plausibleClient.referrers as jest.Mock).mockResolvedValueOnce([
    {
      pageviews: '5',
      source: 'https://stacks.co',
    },
  ]);

  const response = await server.inject({
    method: 'GET',
    url: '/api/analytics/referrers?dateFrom=2022-03-15&storyId=test',
    cookies: {
      'next-auth.session-token': stacksAddress,
    },
  });

  expect(response.statusCode).toBe(200);
  expect(response.json()).toMatchSnapshot();
  expect(plausibleClient.referrers).toBeCalledWith({
    dateFrom: '2022-05-01',
    dateTo: '2022-04-04',
    paths: ['/sigleapp.id.blockstack/test'],
  });
});
