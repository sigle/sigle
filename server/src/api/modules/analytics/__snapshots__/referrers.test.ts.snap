// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Respond with referrers 1`] = `
Array [
  Object {
    "count": 10,
    "domain": "null",
  },
  Object {
    "count": 6,
    "domain": "https://stacks.co",
  },
]
`;

exports[`Respond with referrers for one story 1`] = `
Array [
  Object {
    "count": 5,
    "domain": "https://stacks.co",
  },
]
`;
